internal/executors/http/oauth_header_generator.go

```diff
@@ -1,179 +1,130 @@
 package http
 
 import (
-	"crypto/sha256"
-	"encoding/hex"
 	"fmt"
-	"log"
-	"net/url"
+	"github.com/SAP/remote-work-processor/internal/utils"
+	"time"
 
 	"github.com/SAP/remote-work-processor/internal/executors/http/tls"
 )
 
-const (
-	CACHING_KEY_FORMAT                         string = "tokenUrl=%s&oAuthUser=%s&oAuthPwd=%s&getTokenBody=%s"
-	PASSWORD_GRANT_FORMAT                      string = "grant_type=password&username=%s&password=%s"
-	PASSWORD_CREDENTIALS_FORMAT_WITH_CLIENT_ID string = "grant_type=password&client_id=%s&username=%s&password=%s"
-	CLIENT_CREDENTIALS_FORMAT                  string = "grant_type=client_credentials&client_id=%s&client_secret=%s"
-	REFRESH_TOKEN_FORMAT                       string = "grant_type=refresh_token&refresh_token=%s"
-	REFRESH_TOKEN_FORMAT_WITH_CERT             string = "grant_type=refresh_token&client_id=%s&refresh_token=%s"
-)
-
-func NewOAuthHeaderGenerator(p *HttpRequestParameters) AuthorizationHeaderGenerator {
-	user := p.GetUser()
-	clientId := p.GetClientId()
-	refreshToken := p.GetRefreshToken()
+type OAuthorizationHeaderOption func(*oAuthorizationHeaderGenerator)
 
-	if refreshToken != "" {
-		return refreshTokenGenerator(p)
-	}
+type oAuthorizationHeaderGenerator struct {
+	tokenType          TokenType
+	certAuthentication *tls.CertificateAuthentication
+	authHeader         string
+	cachingKey         string
+	requestStore       map[string]string
+	fetcher            TokenFetcher
+}
 
-	if user != "" && clientId != "" {
-		if p.GetCertificateAuthentication().GetClientCertificate() != "" {
-			return passwordGrantWithClientCertificateGenerator(p)
-		}
+type cachedToken struct {
+	*OAuthToken
+	IssuedAt int64 `json:"timestamp,omitempty"`
+}
 
-		return passwordGrantGenerator(p)
+func NewOAuthorizationHeaderGenerator(tokenType TokenType, tokenUrl string, executor HttpExecutor, requestBody string,
+	opts ...OAuthorizationHeaderOption) CacheableAuthorizationHeaderGenerator {
+	h := &oAuthorizationHeaderGenerator{
+		tokenType: tokenType,
 	}
 
-	if user != "" {
-		return clientCredentialsGenerator(p, user, p.GetPassword())
+	for _, opt := range opts {
+		opt(h)
 	}
 
-	if clientId != "" {
-		return clientCredentialsGenerator(p, clientId, p.GetClientSecret())
-	}
+	h.fetcher = NewOAuthTokenFetcher(
+		withExecutor(executor),
+		withTokenUrl(tokenUrl),
+		withRequestBody(requestBody),
+		withCertificateAuthentication(h.certAuthentication),
+		withAuthHeader(h.authHeader),
+	)
```

Comment:
Are these `with*` arguments all optional? Can we make an OAuth TokenFetcher without an URL for instance? Feels weird to me to have a builder pattern where non-optional fields are in fact optional (compilation wise) as any errors would be discovered on runtime rather than on the much more preferred compile time.

Can we perhaps consider making the always required parameters as concrete parameters of the function and the optional ones (I think f.certAuthentication might be?) to be left as the optionals at the end of the function?

As an example the `NewOAuthorizationHeaderGenerator` seems to be written in such a way.